### DISCUSSION TITLE
Machine memory limit with podman under QEMU on MacOS and Apple Silicon

### DISCUSSION THREAD

**Discussion by chainmail** [containers/podman#19528] [Q&A] [answered]
This is mostly a FYI. I have reported the issue to QEMU (https://gitlab.com/qemu-project/qemu/-/issues/1816). There appears to be a memory limit in the QEMU emulator for podman machine if you want to use more than 64512 Mb RAM.

The error we get is:

`Starting machine "podman-machine-default"
Waiting for VM ...
Error: qemu exited unexpectedly with exit code -1, stderr: qemu-system-aarch64: Error: HV_BAD_ARGUMENT`

We are using Mac Studios (M2 Ultra) to run podman containers which are used to build the aarch64 release of our product. Unfortunately, we are unable to use all the RAM (128Gb). We get the above error when we initialise the podman machine with any number >64512Mb RAM. Obviously, we can run more than one machine, at the same time that complicates our build environment and we prefer to not do that.

The work around is to do something like this:

`podman machine init --cpus="18" --memory="64512" podman-machine-default`

Similarly to reproduce, just initialise with 64513 Mb RAM.

I could not find this anywhere, so I am posting to visualize it for others.

**Comment by pm215** [comment, marked as answer]
To propagate the answer from the QEMU bug report, this is a hardware limitation of the CPU you are running on. These Apple CPUs seem to have an intermediate physical address space (IPA) limit of 36 bits. That means you can't create a VM with so much RAM that the RAM would need to go above the 64GB mark. (You hit the limit at a bit lower than 64GB passed to the VM because RAM doesn't start at 0GB and there is some space used in the VM address space for devices and so on.)

You'll need to use different host hardware if you want to run larger VMs than this, I'm afraid. (Non-Apple AArch64 CPUs generally have an IPA limit of 40 bits or larger.)
